qiskit/transpiler/passes/synthesis/unitary_synthesis.py

```diff
@@ -445,25 +501,59 @@ def _synth_natural_direction(
         wires = None
         if natural_direction in {None, True} and (coupling_map or target):
             if target is not None:
-                cmap = target.build_coupling_map(two_q_gate=decomposer2q.gate.name)
+                matching = {}
+                reverse = {}
+                # Find lowest error matching or reverse decomposer and use that
+                for index, decomposer in enumerate(decomposer2q):
+                    gate_name = getattr(decomposer, "gate_name", decomposer.gate.name)
+                    qubits_tuple = tuple(qubits)
+                    reverse_tuple = (qubits[1], qubits[0])
+                    props_dict = target[gate_name]
+                    if target.instruction_supported(gate_name, qubits_tuple):
+                        if props_dict is None or None in props_dict:
+                            error = 0.0
+                        else:
+                            error = getattr(props_dict[qubits_tuple], "error", 0.0)
+                            if error is None:
+                                error = 0.0
+                        matching[index] = error
+                    # Skip reverse check if we already have matching
+                    elif not matching and target.instruction_supported(gate_name, reverse_tuple):
+                        if props_dict is None or None in props_dict:
+                            error = 0.0
+                        else:
+                            error = getattr(props_dict[reverse_tuple], "error", 0.0)
+                            if error is None:
+                                error = 0.0
+                        reverse[index] = error
+                preferred_direction = None
+                if matching:
+                    preferred_direction = [0, 1]
+                    min_error_index = min(matching, key=matching.get)
+                    decomposer2q = decomposer2q[min_error_index]
+                elif reverse:
+                    preferred_direction = [1, 0]
+                    min_error_index = min(reverse, key=matching.get)
+                    decomposer2q = decomposer2q[min_error_index]
             else:
                 cmap = coupling_map
-            neighbors0 = cmap.neighbors(qubits[0])
-            zero_one = qubits[1] in neighbors0
-            neighbors1 = cmap.neighbors(qubits[1])
-            one_zero = qubits[0] in neighbors1
-            if zero_one and not one_zero:
-                preferred_direction = [0, 1]
-            if one_zero and not zero_one:
-                preferred_direction = [1, 0]
+                neighbors0 = cmap.neighbors(qubits[0])
+                zero_one = qubits[1] in neighbors0
+                neighbors1 = cmap.neighbors(qubits[1])
+                one_zero = qubits[0] in neighbors1
+                if zero_one and not one_zero:
+                    preferred_direction = [0, 1]
+                if one_zero and not zero_one:
+                    preferred_direction = [1, 0]
```

Comment:
This feels logically like it's part of choosing the 2q decomposer - it doesn't depend on the actual gate at all.  Perhaps the weirdnesses surrounding varying return types in those parts could be eased if this logic was moved (and separated out into a couple of helper functions) into `_basis_gate_to_decomposer_2q` instead?

Also, since this PR is potentially adding a lot more work into that function: how expensive is all this to do?  Is it worth (potentially in another PR) deferring the choice of the 1q and 2q decomposers into only the branches that actually ask for them?  Each `DefaultUnitarySynthesis.run` only acts on one gate each time, so we're always certainly wasting at least _some_ effort.

Reply:
I did think about the performance penalty, but was hoping it wasn't too bad in practice (we don't have a lot of `Target` users yet so it hasn't come up), the problem is it's partially a function of each individual call to `run()`. For the target path we have more stuff to evaluate and want to pick the best decomposer basis for the given qubits it's applied to. The only thing I think we can do here is probably cache the best decomposer on an ordered pair of qubits after it's found and fast path the return. Which definitely goes in with a helper function or two to make this easier to work with.

Reply:
I added a cache in: https://github.com/Qiskit/qiskit-terra/pull/7795/commits/37432e5e410794f2f0cc869ed8c67f8f30a5cb34